File: packages/ts/cairn-p2p/src/pairing/link.ts
import { CairnError } from '../errors.js';
import type { PairingPayload, ConnectionHint } from './payload.js';

/** Default URI scheme. */
const DEFAULT_SCHEME = 'cairn';

/**
 * Generate a pairing link URI from a PairingPayload.
 *
 * Format: `cairn://pair?pid=<hex>&nonce=<hex>&pake=<hex>&hints=<comma-separated>&t=<created>&x=<expires>`
 *
 * Note: The spec says pid is base58. We use hex for simplicity since we don't have
 * a base58 dependency. Hints are JSON-encoded and percent-encoded in the query string.
 */
export function generatePairingLink(
  payload: PairingPayload,
  scheme: string = DEFAULT_SCHEME,
): string {
  const pid = bytesToHex(payload.peerId);
  const nonce = bytesToHex(payload.nonce);
  const pake = bytesToHex(payload.pakeCredential);

  let uri = `${scheme}://pair?pid=${pid}&nonce=${nonce}&pake=${pake}`;

  if (payload.hints && payload.hints.length > 0) {
    const hintsStr = payload.hints
      .map(h => `${h.hintType}:${h.value}`)
      .join(',');
    uri += `&hints=${encodeURIComponent(hintsStr)}`;
  }

  uri += `&t=${payload.createdAt}&x=${payload.expiresAt}`;

  return uri;
}

/**
 * Parse and validate a pairing link URI into a PairingPayload.
 */
export function parsePairingLink(
  uri: string,
  scheme: string = DEFAULT_SCHEME,
): PairingPayload {
  // Validate scheme
  if (!uri.startsWith(`${scheme}://pair?`)) {
    throw new CairnError('PAIRING', `invalid pairing link: expected ${scheme}://pair?...`);
  }

  // Extract query string
  const queryStart = uri.indexOf('?');
  if (queryStart === -1) {
    throw new CairnError('PAIRING', 'invalid pairing link: missing query string');
  }

  const params = new URLSearchParams(uri.slice(queryStart + 1));

  // pid (hex)
  const pidHex = params.get('pid');
  if (!pidHex) {
    throw new CairnError('PAIRING', "missing 'pid' parameter in pairing link");
  }
  const peerId = hexToBytes(pidHex);

  // nonce (hex)
  const nonceHex = params.get('nonce');
  if (!nonceHex) {
    throw new CairnError('PAIRING', "missing 'nonce' parameter in pairing link");
  }
  const nonce = hexToBytes(nonceHex);
  if (nonce.length !== 16) {
    throw new CairnError('PAIRING', `nonce must be 16 bytes, got ${nonce.length}`);
  }

  // pake (hex)
  const pakeHex = params.get('pake');
  if (!pakeHex) {
    throw new CairnError('PAIRING', "missing 'pake' parameter in pairing link");
  }
  const pakeCredential = hexToBytes(pakeHex);

  // hints (optional, comma-separated type:value pairs)
  let hints: ConnectionHint[] | undefined;
  const hintsStr = params.get('hints');
  if (hintsStr) {
    hints = hintsStr.split(',').map(part => {
      const colonIdx = part.indexOf(':');
      if (colonIdx === -1) {
        throw new CairnError('PAIRING', `invalid hint format: '${part}'`);
      }
      return {
        hintType: part.slice(0, colonIdx),
        value: part.slice(colonIdx + 1),
      };
    });
  }

  // timestamps
  const createdAt = parseInt(params.get('t') ?? '0', 10) || 0;
  const expiresAt = parseInt(params.get('x') ?? '0', 10) || 0;

  // Validate expiry
  const now = Math.floor(Date.now() / 1000);
  if (now > expiresAt) {
    throw new CairnError('PAIRING', 'pairing link has expired');
  }

  return { peerId, nonce, pakeCredential, hints, createdAt, expiresAt };
}

/** Convert bytes to hex string. */
function bytesToHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map(b => b.toString(16).padStart(2, '0'))
    .join('');
}

/** Convert hex string to bytes. */
function hexToBytes(hex: string): Uint8Array {
  if (hex.length % 2 !== 0) {
    throw new CairnError('PAIRING', `invalid hex string length: ${hex.length}`);
  }
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.slice(i, i + 2), 16);
  }
  return bytes;
}

File: packages/go/cairn-p2p/pairing/link.go
package pairing

import (
	"crypto/rand"
	"encoding/hex"
	"fmt"
	"io"
	"net/url"
	"strings"
	"time"

	"github.com/moukrea/cairn/packages/go/cairn-p2p/crypto"
	"github.com/mr-tron/base58"
)

// LinkPairingData holds the data encoded in a pairing link URI.
type LinkPairingData struct {
	PeerID    [34]byte
	Nonce     [16]byte
	PakeCred  [32]byte
	Hints     []string
	CreatedAt uint64
	ExpiresAt uint64
}

// GeneratePairingLink creates a pairing link URI.
// Returns the pairing data and the URI string.
func GeneratePairingLink(identity *crypto.IdentityKeypair, ttl time.Duration, hints []string) (*LinkPairingData, string, error) {
	var nonce [16]byte
	if _, err := io.ReadFull(rand.Reader, nonce[:]); err != nil {
		return nil, "", fmt.Errorf("nonce generation failed: %w", err)
	}

	var pakeCred [32]byte
	if _, err := io.ReadFull(rand.Reader, pakeCred[:]); err != nil {
		return nil, "", fmt.Errorf("PAKE credential generation failed: %w", err)
	}

	now := uint64(time.Now().Unix())
	data := &LinkPairingData{
		PeerID:    identity.PeerID(),
		Nonce:     nonce,
		PakeCred:  pakeCred,
		Hints:     hints,
		CreatedAt: now,
		ExpiresAt: now + uint64(ttl.Seconds()),
	}

	uri := fmt.Sprintf("cairn://pair?pid=%s&nonce=%s&pake=%s",
		base58.Encode(data.PeerID[:]),
		hex.EncodeToString(nonce[:]),
		hex.EncodeToString(pakeCred[:]),
	)

	if len(hints) > 0 {
		uri += "&hints=" + url.QueryEscape(strings.Join(hints, ","))
	}

	uri += fmt.Sprintf("&t=%d&x=%d", data.CreatedAt, data.ExpiresAt)

	return data, uri, nil
}

// ParsePairingLink parses a cairn:// pairing link URI.
func ParsePairingLink(uri string) (*LinkPairingData, error) {
	u, err := url.Parse(uri)
	if err != nil {
		return nil, fmt.Errorf("invalid URI: %w", err)
	}
	if u.Scheme != "cairn" {
		return nil, fmt.Errorf("expected scheme 'cairn', got '%s'", u.Scheme)
	}
	if u.Host != "pair" {
		return nil, fmt.Errorf("expected host 'pair', got '%s'", u.Host)
	}

	params := u.Query()

	// pid (base58)
	pidStr := params.Get("pid")
	if pidStr == "" {
		return nil, fmt.Errorf("missing 'pid' parameter")
	}
	pidBytes, err := base58.Decode(pidStr)
	if err != nil {
		return nil, fmt.Errorf("invalid pid base58: %w", err)
	}
	if len(pidBytes) != 34 {
		return nil, fmt.Errorf("invalid pid length: got %d, want 34", len(pidBytes))
	}
	var peerID [34]byte
	copy(peerID[:], pidBytes)

	// nonce (hex)
	nonceStr := params.Get("nonce")
	if nonceStr == "" {
		return nil, fmt.Errorf("missing 'nonce' parameter")
	}
	nonceBytes, err := hex.DecodeString(nonceStr)
	if err != nil {
		return nil, fmt.Errorf("invalid hex nonce: %w", err)
	}
	if len(nonceBytes) != 16 {
		return nil, fmt.Errorf("nonce must be 16 bytes, got %d", len(nonceBytes))
	}
	var nonce [16]byte
	copy(nonce[:], nonceBytes)

	// pake (hex)
	pakeStr := params.Get("pake")
	if pakeStr == "" {
		return nil, fmt.Errorf("missing 'pake' parameter")
	}
	pakeBytes, err := hex.DecodeString(pakeStr)
	if err != nil {
		return nil, fmt.Errorf("invalid hex pake: %w", err)
	}
	if len(pakeBytes) != 32 {
		return nil, fmt.Errorf("pake credential must be 32 bytes, got %d", len(pakeBytes))
	}
	var pakeCred [32]byte
	copy(pakeCred[:], pakeBytes)

	// hints (optional, comma-separated)
	var hints []string
	if hintsStr := params.Get("hints"); hintsStr != "" {
		hints = strings.Split(hintsStr, ",")
	}

	// timestamps
	var createdAt, expiresAt uint64
	if tStr := params.Get("t"); tStr != "" {
		fmt.Sscanf(tStr, "%d", &createdAt)
	}
	if xStr := params.Get("x"); xStr != "" {
		fmt.Sscanf(xStr, "%d", &expiresAt)
	}

	return &LinkPairingData{
		PeerID:    peerID,
		Nonce:     nonce,
		PakeCred:  pakeCred,
		Hints:     hints,
		CreatedAt: createdAt,
		ExpiresAt: expiresAt,
	}, nil
}

// IsExpired reports whether the link data has expired.
func (l *LinkPairingData) IsExpired() bool {
	return uint64(time.Now().Unix()) > l.ExpiresAt
}
